runtime/lua/vim/lsp/folding_range.lua

```diff
@@ -0,0 +1,365 @@
+  return request_id
+end
+
+-- MAINTENANCE NOTE:
+-- * `bufstate` requires event hooks to ensure correctness.
+-- * event hooks requires the previous `bufstate` for updates.
+--
+-- In other words, they are interdependent.
+-- Since their creation and destruction are handled manually,
+-- our strategy is to ensure that their lifecycles are always synchronized.
+--
+-- As they are only meaningful for buffers associated with
+-- at least one window that have `foldexpr()` set,
+-- the following three elements are guaranteed
+-- to exist or not exist simultaneously:
+--
+-- * A buffer associated with at least one window where `foldexpr()` is set.
+-- * Its `bufstate`.
+-- * Its event hooks.
+local augroup_setup = api.nvim_create_augroup('vim_lsp_folding_range/setup', {})
+
+--- Setup `bufstate` and its event hooks.
+---@param bufnr integer
+---@return vim.lsp.folding_range.BufState
+local function setup(bufnr)
+  ---@type vim.lsp.folding_range.BufState
+  local bufstate = { row_level = {}, row_text = {}, kind_ranges = {} }
+  if not api.nvim_buf_is_loaded(bufnr) then
+    return bufstate
+  end
+
+  -- Register `bufstate`.
+  bufstates[bufnr] = bufstate
+
+  -- Setup event hooks.
+  api.nvim_buf_attach(bufnr, false, {
+    on_detach = function()
+      bufstates[bufnr] = nil
+    end,
+    on_reload = function()
+      bufstate = { row_level = {}, row_text = {}, kind_ranges = {} }
+      bufstates[bufnr] = bufstate
+      for _, client in
+        ipairs(vim.lsp.get_clients({ bufnr = bufnr, method = ms.textDocument_foldingRange }))
+      do
+        refresh(bufnr, client)
+      end
+    end,
+    --- When modifying the buffer, new foldinfo is scheduled to be updated,
+    --- synchronize the changed rows with their previous foldlevels as possible,
+    --- before new foldlevels is updated.
+    ---@param start_row integer
+    ---@param old_row integer
+    ---@param new_row integer
+    on_bytes = function(_, _, _, start_row, _, _, old_row, _, _, new_row, _, _)
+      if bufstate == nil then
+        return true
+      end
+      local row = new_row - old_row
+      local row_level = bufstate.row_level
+      if vim.tbl_isempty(row_level) then
+        return
+      end
+      if row > 0 then
+        vim._list_insert(bufstate.row_level, start_row, start_row + math.abs(row) - 1, { -1 })
+      elseif row < 0 then
+        vim._list_remove(bufstate.row_level, start_row, start_row + math.abs(row) - 1)
+      end
+    end,
+  })
+  api.nvim_create_autocmd('LspDetach', {
+    group = augroup_setup,
+    buffer = bufnr,
+    callback = function()
+      local clients = vim.lsp.get_clients({ bufnr = bufnr, method = ms.textDocument_foldingRange })
+      if #clients == 0 then
+        for _, client in ipairs(clients) do
+          refresh(bufnr, client)
+        end
+      else
+        bufstate = { row_level = {}, row_text = {}, kind_ranges = {} }
+        bufstates[bufnr] = bufstate
```

Comment:
I don't think you should be registering bufstate here.

Reply:
This is not registering but resetting bufstate. The assignment `bufstates[bufnr] = bufstate` needs to be done because `bufstates[bufnr]` still points to the previously constructed `bufstate`, not the new one.